src/pbkdf2.rs

```diff
@@ -1,10 +1,94 @@
 
 use bitcoin_hashes::{hmac, sha512, Hash, HashEngine};
 
+const SALT_PREFIX: &'static str = "mnemonic";
+
+/// Calculate the binary size of the mnemonic.
+fn mnemonic_byte_len(mnemonic: &[&'static str]) -> usize {
+	let mut len = 0;
+	for i in 0..mnemonic.len() {
+		let word = &mnemonic[i];
+		if word.is_empty() {
+			break;
+		}
+		if i > 0 {
+			len += 1;
+		}
+		len += word.len();
+	}
+	len
+}
+
+/// Wrote the mnemonic in binary form into the hash engine.
+fn mnemonic_write_into(mnemonic: &[&'static str], engine: &mut sha512::HashEngine) {
+	for i in 0..mnemonic.len() {
+		let word = &mnemonic[i];
+		if word.is_empty() {
+			break;
+		}
+		if i > 0 {
+			engine.input(" ".as_bytes());
+		}
+		engine.input(word.as_bytes());
+	}
+}
+
+/// Create an HMAC engine from the passphrase.
+/// We need a special method because we can't allocate a new byte
+/// vector for the entire serialized mnemonic.
+fn create_hmac_engine(mnemonic: &[&'static str]) -> hmac::HmacEngine<sha512::Hash> {
+	// Inner code is borrowed from the bitcoin_hashes::hmac::HmacEngine::new method.
+	let mut ipad = [0x36u8; 128];
+	let mut opad = [0x5cu8; 128];
+	let mut iengine = sha512::Hash::engine();
+	let mut oengine = sha512::Hash::engine();
+
+	if mnemonic_byte_len(mnemonic) > sha512::HashEngine::BLOCK_SIZE {
+		let hash = {
+			let mut engine = sha512::Hash::engine();
+			mnemonic_write_into(mnemonic, &mut engine);
+			sha512::Hash::from_engine(engine)
+		};
+
+		for (b_i, b_h) in ipad.iter_mut().zip(&hash[..]) {
+			*b_i ^= *b_h;
+		}
+		for (b_o, b_h) in opad.iter_mut().zip(&hash[..]) {
+			*b_o ^= *b_h;
+		}
+	} else {
+		// First modify the first elements from the prefix.
+		let mut cursor = 0;
+		for i in 0..mnemonic.len() {
+			let word = &mnemonic[i];
+			if word.is_empty() {
+				break;
+			}
+			if i > 0 {
+				ipad[cursor] ^= ' ' as u8;
+				opad[cursor] ^= ' ' as u8;
+				cursor += 1;
+			}
+			for (b_i, b_h) in ipad.iter_mut().skip(cursor).zip(word.as_bytes()) {
+				*b_i ^= *b_h;
+			}
+			for (b_o, b_h) in opad.iter_mut().skip(cursor).zip(word.as_bytes()) {
+				*b_o ^= *b_h;
+			}
+			cursor += word.len();
+			assert!(cursor <= sha512::HashEngine::BLOCK_SIZE, "mnemonic_byte_len is broken");
+		}
+	};
+
+	iengine.input(&ipad[..sha512::HashEngine::BLOCK_SIZE]);
+	oengine.input(&opad[..sha512::HashEngine::BLOCK_SIZE]);
+	hmac::HmacEngine::from_inner_engines(iengine, oengine)
+}
+
 // Method borrowed from rust-bitcoin's endian module.
 #[inline]
 fn u32_to_array_be(val: u32) -> [u8; 4] {
-	debug_assert_eq!(::std::mem::size_of::<u32>(), 4); // size_of isn't a constfn in 1.22
+	debug_assert_eq!(::core::mem::size_of::<u32>(), 4); // size_of isn't a constfn in 1.22
```

Comment:
That's a very silly assert imo xD why do you even care about the size here? Like, what would change if the int was like 4GB in size because it's encoded in unary ^^